File: lib/src/apps/quran/core/entities/quran_entities.dart
import 'package:equatable/equatable.dart';

/// كيان السورة القرآنية
class SurahEntity extends Equatable {
  final int id;
  final int? juzId;
  final int? pageId;
  final int? fallOrder;
  final int? moshafOrder;
  final int? ayahsCount;
  final int? fallPlace; // 0=مكية, 1=مدنية
  final String title;
  final String? names;
  final String? info;
  final String symbol;
  final int? startAyahOrder;

  const SurahEntity({
    required this.id,
    this.juzId,
    this.pageId,
    this.fallOrder,
    this.moshafOrder,
    this.ayahsCount,
    this.fallPlace,
    required this.title,
    this.names,
    this.info,
    required this.symbol,
    this.startAyahOrder,
  });

  bool get isMakki => fallPlace == 0;
  bool get isMadani => fallPlace == 1;

  @override
  List<Object?> get props => [
        id,
        juzId,
        pageId,
        fallOrder,
        moshafOrder,
        ayahsCount,
        fallPlace,
        title,
        names,
        info,
        symbol,
        startAyahOrder,
      ];
}

/// كيان الآية القرآنية
class AyahEntity extends Equatable {
  final int id;
  final int? surahId;
  final int? juzId;
  final int? hizbId;
  final int? pageId;
  final int? hizbNumber;
  final int? ayahOrder;
  final int? ayahNumber;
  final bool isStartHizb;
  final bool isStartJuz;
  final bool hasProstration;
  final int? prostrationType;
  final String? plainText;
  final String? quranText;
  final String? stem;

  const AyahEntity({
    required this.id,
    this.surahId,
    this.juzId,
    this.hizbId,
    this.pageId,
    this.hizbNumber,
    this.ayahOrder,
    this.ayahNumber,
    this.isStartHizb = false,
    this.isStartJuz = false,
    this.hasProstration = false,
    this.prostrationType,
    this.plainText,
    this.quranText,
    this.stem,
  });

  @override
  List<Object?> get props => [
        id,
        surahId,
        juzId,
        hizbId,
        pageId,
        hizbNumber,
        ayahOrder,
        ayahNumber,
        isStartHizb,
        isStartJuz,
        hasProstration,
        prostrationType,
        plainText,
        quranText,
        stem,
      ];
}

/// كيان صفحة المصحف
class PageEntity extends Equatable {
  final int id;
  final int? startSurahId;
  final int? startJuzId;
  final int? startHizbId;
  final int? hizbNumber;
  final int? pageNumber;
  final int? startAyahNumber;
  final int? startAyahOrder;

  const PageEntity({
    required this.id,
    this.startSurahId,
    this.startJuzId,
    this.startHizbId,
    this.hizbNumber,
    this.pageNumber,
    this.startAyahNumber,
    this.startAyahOrder,
  });

  @override
  List<Object?> get props => [
        id,
        startSurahId,
        startJuzId,
        startHizbId,
        hizbNumber,
        pageNumber,
        startAyahNumber,
        startAyahOrder,
      ];
}

/// كيان الجزء
class JuzEntity extends Equatable {
  final int id;
  final int? startSurahId;
  final int? startPageId;
  final int? juzNumber;
  final int? startAyahNumber;
  final String? firstAyahs;
  final int? startAyahOrder;

  const JuzEntity({
    required this.id,
    this.startSurahId,
    this.startPageId,
    this.juzNumber,
    this.startAyahNumber,
    this.firstAyahs,
    this.startAyahOrder,
  });

  @override
  List<Object?> get props => [
        id,
        startSurahId,
        startPageId,
        juzNumber,
        startAyahNumber,
        firstAyahs,
        startAyahOrder,
      ];
}

/// كيان الحزب
class HizbEntity extends Equatable {
  final int id;
  final int? juzId;
  final int? surahId;
  final int? startPageId;
  final int? hizbNumber;
  final int? segmentOrder;
  final int? ayahNumber;
  final String? firstAyahs;
  final int? startAyahOrder;
  final int? approvalStatus;

  const HizbEntity({
    required this.id,
    this.juzId,
    this.surahId,
    this.startPageId,
    this.hizbNumber,
    this.segmentOrder,
    this.ayahNumber,
    this.firstAyahs,
    this.startAyahOrder,
    this.approvalStatus,
  });

  String get segmentName {
    switch (segmentOrder) {
      case 1:
        return 'الربع الأول';
      case 2:
        return 'النصف';
      case 3:
        return 'الثلث';
      case 4:
        return 'الربع الأخير';
      default:
        return '';
    }
  }

  @override
  List<Object?> get props => [
        id,
        juzId,
        surahId,
        startPageId,
        hizbNumber,
        segmentOrder,
        ayahNumber,
        firstAyahs,
        startAyahOrder,
        approvalStatus,
      ];
}

File: huda_allah_system/apps/backend/src/database/entities/quran.entities.ts
import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, OneToMany, JoinColumn } from 'typeorm';

@Entity('quran_surah')
export class Surah {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'juz_id', nullable: true, type: 'bigint' })
  juzId: number;

  @Column({ name: 'page_id', nullable: true, type: 'bigint' })
  pageId: number;

  @Column({ name: 'fall_order', nullable: true, type: 'int', unsigned: true })
  fallOrder: number;

  @Column({ name: 'moshaf_order', nullable: true, type: 'int', unsigned: true })
  moshafOrder: number;

  @Column({ name: 'ayahs_count', nullable: true, type: 'int', unsigned: true })
  ayahsCount: number;

  @Column({ name: 'fall_place', nullable: true })
  fallPlace: number; // 0 = مكية, 1 = مدنية

  @Column({ length: 50 })
  title: string;

  @Column({ nullable: true, type: 'text' })
  names: string;

  @Column({ nullable: true, type: 'text' })
  info: string;

  @Column({ length: 5 })
  symbol: string;

  @Column({ name: 'start_ayah_order', nullable: true, type: 'int', unsigned: true })
  startAyahOrder: number;

  // Relations
  @OneToMany(() => Ayah, ayah => ayah.surah)
  ayahs: Ayah[];

  @ManyToOne(() => Juz, juz => juz.surahs)
  @JoinColumn({ name: 'juz_id' })
  juz: Juz;

  @ManyToOne(() => Page, page => page.surahs)
  @JoinColumn({ name: 'page_id' })
  page: Page;

  // Computed properties
  get isMakki(): boolean {
    return this.fallPlace === 0;
  }

  get isMadani(): boolean {
    return this.fallPlace === 1;
  }
}

@Entity('quran_ayah')
export class Ayah {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'surah_id', nullable: true, type: 'bigint' })
  surahId: number;

  @Column({ name: 'juz_id', nullable: true, type: 'bigint' })
  juzId: number;

  @Column({ name: 'hizb_id', nullable: true, type: 'bigint' })
  hizbId: number;

  @Column({ name: 'page_id', nullable: true, type: 'bigint' })
  pageId: number;

  @Column({ name: 'hizb_number', nullable: true, type: 'int', unsigned: true })
  hizbNumber: number;

  @Column({ name: 'ayah_order', nullable: true, type: 'int', unsigned: true })
  ayahOrder: number;

  @Column({ name: 'ayah_number', nullable: true, type: 'int', unsigned: true })
  ayahNumber: number;

  @Column({ name: 'is_start_hizb', default: false })
  isStartHizb: boolean;

  @Column({ name: 'is_start_juz', default: false })
  isStartJuz: boolean;

  @Column({ name: 'has_prostration', default: false })
  hasProstration: boolean;

  @Column({ name: 'prostration_type', nullable: true, type: 'int', unsigned: true })
  prostrationType: number;

  @Column({ name: 'plain_text', nullable: true, type: 'text' })
  plainText: string;

  @Column({ name: 'quran_text', nullable: true, type: 'text' })
  quranText: string;

  @Column({ nullable: true, type: 'text' })
  stem: string;

  // Relations
  @ManyToOne(() => Surah, surah => surah.ayahs)
  @JoinColumn({ name: 'surah_id' })
  surah: Surah;

  @ManyToOne(() => Juz, juz => juz.ayahs)
  @JoinColumn({ name: 'juz_id' })
  juz: Juz;

  @ManyToOne(() => Hizb, hizb => hizb.ayahs)
  @JoinColumn({ name: 'hizb_id' })
  hizb: Hizb;

  @ManyToOne(() => Page, page => page.ayahs)
  @JoinColumn({ name: 'page_id' })
  page: Page;
}

@Entity('quran_page')
export class Page {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'start_surah_id', nullable: true, type: 'bigint' })
  startSurahId: number;

  @Column({ name: 'start_juz_id', nullable: true, type: 'bigint' })
  startJuzId: number;

  @Column({ name: 'start_hizb_id', nullable: true, type: 'bigint' })
  startHizbId: number;

  @Column({ name: 'hizb_number', nullable: true, type: 'int', unsigned: true })
  hizbNumber: number;

  @Column({ name: 'page_number', nullable: true, type: 'int', unsigned: true })
  pageNumber: number;

  @Column({ name: 'start_ayah_number', nullable: true, type: 'int', unsigned: true })
  startAyahNumber: number;

  @Column({ name: 'start_ayah_order', nullable: true, type: 'int', unsigned: true })
  startAyahOrder: number;

  // Relations
  @OneToMany(() => Surah, surah => surah.page)
  surahs: Surah[];

  @OneToMany(() => Ayah, ayah => ayah.page)
  ayahs: Ayah[];
}

@Entity('quran_juz')
export class Juz {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'start_surah_id', nullable: true, type: 'bigint' })
  startSurahId: number;

  @Column({ name: 'start_page_id', nullable: true, type: 'bigint' })
  startPageId: number;

  @Column({ name: 'juz_number', nullable: true, type: 'int', unsigned: true })
  juzNumber: number;

  @Column({ name: 'start_ayah_number', nullable: true, type: 'int', unsigned: true })
  startAyahNumber: number;

  @Column({ name: 'first_ayahs', nullable: true, type: 'text' })
  firstAyahs: string;

  @Column({ name: 'start_ayah_order', nullable: true, type: 'int', unsigned: true })
  startAyahOrder: number;

  // Relations
  @OneToMany(() => Surah, surah => surah.juz)
  surahs: Surah[];

  @OneToMany(() => Ayah, ayah => ayah.juz)
  ayahs: Ayah[];
}

@Entity('quran_hizb')
export class Hizb {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'juz_id', nullable: true, type: 'bigint' })
  juzId: number;

  @Column({ name: 'surah_id', nullable: true, type: 'bigint' })
  surahId: number;

  @Column({ name: 'start_page_id', nullable: true, type: 'bigint' })
  startPageId: number;

  @Column({ name: 'hizb_number', nullable: true, type: 'int', unsigned: true })
  hizbNumber: number;

  @Column({ name: 'segment_order', nullable: true, type: 'int', unsigned: true })
  segmentOrder: number;

  @Column({ name: 'ayah_number', nullable: true, type: 'int', unsigned: true })
  ayahNumber: number;

  @Column({ name: 'first_ayahs', nullable: true, type: 'text' })
  firstAyahs: string;

  @Column({ name: 'start_ayah_order', nullable: true, type: 'int', unsigned: true })
  startAyahOrder: number;

  @Column({ name: 'approval_status', nullable: true })
  approvalStatus: number;

  // Relations
  @OneToMany(() => Ayah, ayah => ayah.hizb)
  ayahs: Ayah[];

  // Computed
  get segmentName(): string {
    const names = ['الربع الأول', 'النصف', 'الثلث', 'الربع الأخير'];
    return names[this.segmentOrder - 1] || '';
  }
}
